chains/ethereum/forwarder_client.go

```diff
@@ -0,0 +1,219 @@
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+	"sync"
+
+	"github.com/ChainSafe/chainbridge-utils/crypto/secp256k1"
+	eth "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	signer "github.com/ethereum/go-ethereum/signer/core"
+)
+
+const ForwarderAbi = "[{\"inputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"constructor\"},{\"inputs\":[{\"components\":[{\"internalType\":\"address\",\"name\":\"from\",\"type\":\"address\"},{\"internalType\":\"address\",\"name\":\"to\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"value\",\"type\":\"uint256\"},{\"internalType\":\"uint256\",\"name\":\"gas\",\"type\":\"uint256\"},{\"internalType\":\"uint256\",\"name\":\"nonce\",\"type\":\"uint256\"},{\"internalType\":\"bytes\",\"name\":\"data\",\"type\":\"bytes\"}],\"internalType\":\"struct MinimalForwarder.ForwardRequest\",\"name\":\"req\",\"type\":\"tuple\"},{\"internalType\":\"bytes\",\"name\":\"signature\",\"type\":\"bytes\"}],\"name\":\"execute\",\"outputs\":[{\"internalType\":\"bool\",\"name\":\"\",\"type\":\"bool\"},{\"internalType\":\"bytes\",\"name\":\"\",\"type\":\"bytes\"}],\"stateMutability\":\"payable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"from\",\"type\":\"address\"}],\"name\":\"getNonce\",\"outputs\":[{\"internalType\":\"uint256\",\"name\":\"\",\"type\":\"uint256\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[{\"components\":[{\"internalType\":\"address\",\"name\":\"from\",\"type\":\"address\"},{\"internalType\":\"address\",\"name\":\"to\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"value\",\"type\":\"uint256\"},{\"internalType\":\"uint256\",\"name\":\"gas\",\"type\":\"uint256\"},{\"internalType\":\"uint256\",\"name\":\"nonce\",\"type\":\"uint256\"},{\"internalType\":\"bytes\",\"name\":\"data\",\"type\":\"bytes\"}],\"internalType\":\"struct MinimalForwarder.ForwardRequest\",\"name\":\"req\",\"type\":\"tuple\"},{\"internalType\":\"bytes\",\"name\":\"signature\",\"type\":\"bytes\"}],\"name\":\"verify\",\"outputs\":[{\"internalType\":\"bool\",\"name\":\"\",\"type\":\"bool\"}],\"stateMutability\":\"view\",\"type\":\"function\"}]"
+
+// Keeps track of forwarder info include the state of the current
+// nonces
+type ForwarderClient struct {
+	client             *ethclient.Client
+	forwarderAddress   common.Address
+	forwarderAbi       abi.ABI
+	forwarderNonceLock sync.Mutex
+	forwarderNonce     *big.Int
+	fromAddress        common.Address
+}
+
+func NewForwarderClient(
+	client *ethclient.Client,
+	forwarderAddress common.Address,
+	fromAddress common.Address,
+) (*ForwarderClient, error) {
+	forwarderAbi, err := abi.JSON(strings.NewReader(ForwarderAbi))
+	if err != nil {
+		return nil, err
+	}
```

Comment:
This is fine as-is, but you might also consider moving this initialization into a package-level `var` at the top of this file and panicking if it fails. Because the value you're parsing is a hardcode, there's really no reason you should ever ship this in a way such that this line fails. By moving this into a package-level `var` you get to _not_ return an error from this constructor and fail on startup if your hardcode is wrong.

Reply:
I added the panic in the constructor, was that what you expected? I couldnt see a way to do any conditional logic at the package level.

I also couldnt figure out how to make the value const - which I thought would be better

Reply:
You're right, you can't do this in a var declaration, not sure what I was thinking. But still, what you've done by panicking in the constructor is an improvement!

This is kind of an edge case, it's not normally a good thing to panic in Go, but in this case because you're dealing with a hardcode as I described, I think it's better to panic and simplify the API of this function.

Reply:
WAIT I REMEMBER HOW TO DO THIS! An `init` function! I'll send you a go playground in a second...

Reply:
You could do this, which would get you panic-on-startup if your hardcode is wrong: https://play.golang.org/p/Tizj1P2COJU

https://golang.org/doc/effective_go#init

You can't declare it as a const unfortunately, because the value of `forwarderAbi` is computed at runtime.